// getstream/lib/lib/files.js
var fetch = require('@stream-io/cross-fetch');

var Headers = require('@stream-io/cross-fetch').Headers;

var utils = require('./utils');

var FormData = require('form-data');

var StreamFileStore = function StreamFileStore() {
  this.initialize.apply(this, arguments);
};

StreamFileStore.prototype = {
  initialize: function initialize(client, token) {
    this.client = client;
    this.token = token;
  },
  upload: function upload(uri, name) {
    var data = new FormData();
    var fileField;

    if (utils.isReadableStream(uri)) {
      fileField = uri;
    } else {
      fileField = {
        uri: uri,
        type: 'application/octet-stream',
        name: name || uri.split('/').reverse()[0]
      };
    }

    data.append('file', fileField);
    return fetch("".concat(this.client.enrichUrl('files/'), "?api_key=").concat(this.client.apiKey), {
      method: 'post',
      body: data,
      headers: new Headers({
        Authorization: this.token
      })
    }).then(function (r) {
      return r.json();
    });
  },
  delete: function _delete(uri) {
    return this.client.delete({
      url: "files/".concat(uri),
      signature: this.token
    });
  }
};
module.exports = StreamFileStore;
