// getstream/lib/lib/images.js
var _extends = require("@babel/runtime/helpers/extends");

var fetch = require('@stream-io/cross-fetch');

var Headers = require('@stream-io/cross-fetch').Headers;

var FormData = require('form-data');

var utils = require('./utils');

var StreamImageStore = function StreamImageStore() {
  this.initialize.apply(this, arguments);
};

StreamImageStore.prototype = {
  initialize: function initialize(client, token) {
    this.client = client;
    this.token = token;
  },
  upload: function upload(uri, name) {
    var data = new FormData();
    var fileField;

    if (utils.isReadableStream(uri)) {
      fileField = uri;
    } else {
      fileField = {
        uri: uri,
        type: 'application/octet-stream',
        name: name || uri.split('/').reverse()[0]
      };
    }

    data.append('file', fileField);
    return fetch("".concat(this.client.enrichUrl('images/'), "?api_key=").concat(this.client.apiKey), {
      method: 'post',
      body: data,
      headers: new Headers({
        Authorization: this.token
      })
    }).then(function (r) {
      return r.json();
    });
  },
  delete: function _delete(uri) {
    return this.client.delete({
      url: "images/".concat(uri),
      signature: this.token
    });
  },
  process: function process(uri, options) {
    var params = _extends(options);

    if (Array.isArray(params.crop)) {
      params.crop = params.crop.join(',');
    }

    return this.client.get({
      url: "images/".concat(uri),
      qs: params,
      signature: this.token
    });
  },
  thumbmail: function thumbmail(uri, w, h) {
    var _ref = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {
      crop: 'center',
      resize: 'clip'
    },
        crop = _ref.crop,
        resize = _ref.resize;

    return this.process(uri, {
      w: w,
      h: h,
      crop: crop,
      resize: resize
    });
  }
};
module.exports = StreamImageStore;
