// getstream/lib/lib/user.js
var StreamUser = function StreamUser() {
  this.initialize.apply(this, arguments);
};

StreamUser.prototype = {
  initialize: function initialize(client, userId, userAuthToken) {
    /**
     * Initialize a user session object
     * @method intialize
     * @memberof StreamUser.prototype
     * @param {StreamClient} client Stream client this collection is constructed from
     * @param {string} userId The ID of the user
     * @param {string} token JWT token
     * @example new StreamUser(client, "123", "eyJhbGciOiJIUzI1...")
     */
    this.client = client;
    this.id = userId;
    this.data = undefined;
    this.full = undefined;
    this.token = userAuthToken;
    this.url = 'user/' + this.id + '/';
  },
  _streamRef: function _streamRef() {
    return "SU:".concat(this.id);
  },
  get: function get(options, callback) {
    var _this = this;

    return this.client.get({
      url: this.url,
      signature: this.token,
      qs: options
    }).then(function (response) {
      _this.full = response;
      _this.data = response.data;

      if (callback) {
        callback(response);
      }

      return response;
    });
  },
  _chooseData: function _chooseData(data) {
    if (data !== undefined) {
      return data;
    }

    if (this.data !== undefined) {
      return this.data;
    }

    return {};
  },
  create: function create(data, options, callback) {
    var _this2 = this;

    return this.client.post({
      url: 'user/',
      body: {
        id: this.id,
        data: this._chooseData(data)
      },
      qs: options,
      signature: this.token
    }).then(function (response) {
      _this2.full = response;
      _this2.data = response.data;

      if (callback) {
        callback(response);
      }

      return response;
    });
  },
  update: function update(data, callback) {
    var _this3 = this;

    return this.client.put({
      url: this.url,
      body: {
        data: this._chooseData(data)
      },
      signature: this.token
    }).then(function (response) {
      _this3.full = response;
      _this3.data = response.data;

      if (callback) {
        callback(response);
      }

      return response;
    });
  },
  getOrCreate: function getOrCreate(data, callback) {
    return this.create(data, {
      get_or_create: true
    }, callback);
  },
  profile: function profile(callback) {
    return this.get({
      with_follow_counts: true
    }, callback);
  }
};
module.exports = StreamUser;
